numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -322,94 +322,99 @@ PyArray_BoolConverter(PyObject *object, npy_bool *val)
-    str = PyBytes_AsString(obj);
-    if (!str) {
-        Py_XDECREF(tmp);
+    else if (PyUnicode_Check(object)) {
+        str_object = object;
+        Py_INCREF(str_object);
+    }
+    else {
+        PyErr_Format(PyExc_TypeError,
+            "%s must be str, not %s", name, Py_TYPE(object)->tp_name);
         return NPY_FAIL;
     }
-    if (strlen(str) < 1) {
-        PyErr_SetString(PyExc_ValueError,
-                        "Byteorder string must be at least length 1");
-        Py_XDECREF(tmp);
+
+    Py_ssize_t length;
+    char const *str = PyUnicode_AsUTF8AndSize(str_object, &length);
+    if (str == NULL) {
+        Py_DECREF(str_object);
         return NPY_FAIL;
     }
-    *endian = str[0];
-    if (str[0] != NPY_BIG && str[0] != NPY_LITTLE
-        && str[0] != NPY_NATIVE && str[0] != NPY_IGNORE) {
-        if (str[0] == 'b' || str[0] == 'B') {
-            *endian = NPY_BIG;
-        }
-        else if (str[0] == 'l' || str[0] == 'L') {
-            *endian = NPY_LITTLE;
-        }
-        else if (str[0] == 'n' || str[0] == 'N') {
-            *endian = NPY_NATIVE;
-        }
-        else if (str[0] == 'i' || str[0] == 'I') {
-            *endian = NPY_IGNORE;
-        }
-        else if (str[0] == 's' || str[0] == 'S') {
-            *endian = NPY_SWAP;
-        }
-        else {
+
+    int ret = str_func(str, length, out);
+    Py_DECREF(str_object);
+    if (ret < 0) {
             PyErr_Format(PyExc_ValueError,
-                         "%s is an unrecognized byteorder",
-                         str);
-            Py_XDECREF(tmp);
-            return NPY_FAIL;
-        }
+                "%s %s (got %R)", name, message, object);
+        return NPY_FAIL;
     }
-    Py_XDECREF(tmp);
     return NPY_SUCCEED;
 }
 
-/*NUMPY_API
- * Convert object to sort kind
- */
-NPY_NO_EXPORT int
-PyArray_SortkindConverter(PyObject *obj, NPY_SORTKIND *sortkind)
+static int byteorder_parser(char const *str, Py_ssize_t length, void *data)
 {
-    char *str;
-    PyObject *tmp = NULL;
+    char *endian = (char *)data;
 
-    if (obj == Py_None) {
-        *sortkind = NPY_QUICKSORT;
-        return NPY_SUCCEED;
+    if (length < 1) {
+        return -1;
     }
-
-    if (PyUnicode_Check(obj)) {
-        obj = tmp = PyUnicode_AsASCIIString(obj);
-        if (obj == NULL) {
-            return NPY_FAIL;
-        }
+    else if (str[0] == NPY_BIG || str[0] == NPY_LITTLE
+        || str[0] == NPY_NATIVE || str[0] == NPY_IGNORE) {
+        *endian = str[0];
+    }
+    else if (str[0] == 'b' || str[0] == 'B') {
+        *endian = NPY_BIG;
+    }
+    else if (str[0] == 'l' || str[0] == 'L') {
+        *endian = NPY_LITTLE;
     }
+    else if (str[0] == 'n' || str[0] == 'N') {
+        *endian = NPY_NATIVE;
+    }
+    else if (str[0] == 'i' || str[0] == 'I') {
+        *endian = NPY_IGNORE;
+    }
+    else if (str[0] == 's' || str[0] == 'S') {
+        *endian = NPY_SWAP;
+    }
+    else {
+        return -1;
+    }
+    return 0;
```

Comment:
```suggestion
    return -1;
```
or even assert if you want. Not that it matters too much.

Reply:
Nvm, this one is reachable, since you have no returns elsewhere... Maybe I was just seeing things yesterday...

Reply:
Yeah, I guess I like the immediate return pattern if nothing else happens, and that made me see things...

Reply:
I can put a `return 0` in every branch if you'd prefer.

Reply:
Nah, I don't care enough either way.

Reply:
Gone ahead and did it anyway, I agree it can be sort of hard to read as written.